activesupport/test/xml_mini/common.rb

```diff
@@ -0,0 +1,229 @@
+module CommonXMLMiniAdapterTest
```

Comment:
Since all but JDOM engine tests are of the form:

```ruby
begin
  require "libxml"
rescue LoadError
  # Skip libxml tests
else
  require "abstract_unit"
  require "active_support/xml_mini"
  require "active_support/core_ext/hash/conversions"
  require_relative "./common"

  class LibxmlEngineTest < ActiveSupport::TestCase
  end
end
```

I think I'd abstract this as an custom `TestCase` instead:

```ruby
# xml_mini/engine_test.rb
require "abstract_unit"
require "active_support/xml_mini"
require "active_support/core_ext/hash/conversions"

class XMLMiniEngineTest < ActiveSupport::TestCase
  def self.run_with_gem(gem_name)
    require gem_name
    yield
  rescue LoadError # Skip tests
  end

  setup do
    @old_engine, XMLMini.backend = XMLMini.backend, engine
  end
  
  teardown do
    XMLMini.backend = @old_engine
  end
  
  # All the standard tests...
end

# xml_mini/libxml_engine_test.rb
require_relative 'xml_mini_engine_test'

XMLMiniEngineTest.skip_unless_gem("libxml") do
  class LibxmlEngineTest < XMLMiniEngineTest
    private
      # I dislike the `adapter_name` name since the nomenclature is `engine` :)
      def engine
        "LibXML"
      end
      
      def expansion_error_attack
        # ...
      end
  end
end
```